feat(vulnerabilities): set, maintain and expose vulnerabilitySeverity for templated fields

setchy opened the pull request:
## Changes

Set `vulnerabilitySeverity` for package rules, maintain the highest rating for the update group and expose for use in any template-eligible fields, such as `commitMessageSuffix`.

Demonstration repository: https://github.com/setchy/renovate-demo-osv-severity

## Context

https://github.com/renovatebot/renovate/discussions/21937

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

viceice commented on lib/config/options/index.ts:
```diff
@@ -1838,6 +1839,14 @@ const options: RenovateOptions[] = [
     advancedUse: true,
   },
+  {
+    name: 'includeVulnerabilitySeverity',
```
i would prefer to avoid this option and instead allow the user to override the template.

@rarkins WDYT?

rarkins replied:
Yes, I'd prefer to make it a preset to opt into

setchy replied:
I originally tried overriding `commitMessageSuffix` using a packageRule (ie: without any `includeVulnerabilitySeverity` flag).

However from my local testing I wasn't able to override the suffix - it would always stay as `[SECURITY]`  🤷 

I imagine the same would be true if I moved this to an opt-in preset

Any tips?

viceice replied:
you probably need to move it to the global vulnerability alert config object, because tht is applied last

secustor commented on lib/workers/repository/process/vulnerabilities.ts:
```diff
@@ -454,12 +454,16 @@ export class Vulnerabilities {
       `Setting allowed version ${fixedVersion} to fix vulnerability ${vulnerability.id} in ${packageName} ${depVersion}`
     );
+
+    const severityLevel = vulnerability.database_specific?.severity as string;
```
Rather than `database_specific` which changes based source, use the general `vulnerability.severity`

setchy replied:
I chose `vulnerability.database_specific?.severity` so that it was consistent with what is used within the prBody https://github.com/renovatebot/renovate/blob/c34ebaa0ea1650c1c8fcc4ba20f78b8b25cfa130/lib/workers/repository/process/vulnerabilities.ts#L533-L537

I did look into `vulnerability.severity`, but it doesn't look suitable.  It is sometimes not present at all (ie: when testing `express-cart` vuls), or when it is (ie: testing `karma` vuls) it is an array containing a score and type (no severity)

Churro replied:
This works only for GitHub security advisories that provide no CVSS base score and is meant as a last resort / fallback instead of showing "Unknown".

setchy replied:
@Churro - ah ok, got it.  i'll rework the logic

setchy replied:
Logic extracted to `extractSeverityDetails` so it can be reused.

setchy commented on lib/config/types.ts:
```diff
@@ -309,6 +309,7 @@ export interface PackageRule
     UpdateConfig,
     Record<string, unknown> {
   description?: string | string[];
+  isVulnerabilityAlert?: boolean;
```
This was a missing type for existing logic - see https://github.com/renovatebot/renovate/blob/c34ebaa0ea1650c1c8fcc4ba20f78b8b25cfa130/lib/workers/repository/process/vulnerabilities.ts#L465

setchy commented:
Update to draft PR pushed that addresses https://github.com/renovatebot/renovate/discussions/21937#discussioncomment-5785817 from @Churro, albeit in a somewhat hacky method

setchy commented:
An alternate name to `highestVulnerabilitySeverity` might be `aggregateVulnerabilitySeverity`.  Also open to other suggestions

Churro commented:
> Update to draft PR pushed that addresses [#21937 (comment)](https://github.com/renovatebot/renovate/discussions/21937#discussioncomment-5785817) from @Churro, albeit in a somewhat hacky method

Probably that was a misunderstanding but I didn't intend to recommend the implementation a hacky solution in my comment. I'm convinced there is a more elegant way to implement this in renovate but, personally, I'd rather spend longer tinkering with a nicer solution than implementing this with a crowbar.

setchy commented:
> > Update to draft PR pushed that addresses [#21937 (comment)](https://github.com/renovatebot/renovate/discussions/21937#discussioncomment-5785817) from @Churro, albeit in a somewhat hacky method
> 
> Probably that was a misunderstanding but I didn't intend to recommend the implementation a hacky solution in my comment. I'm convinced there is a more elegant way to implement this in renovate but, personally, I'd rather spend longer tinkering with a nicer solution than implementing this with a crowbar.

I was only referring to the multiple vulnerabilities piece you pointed out.  not suggesting you recommended this method of implementation.  as always, absolutely happy to iterate on these PRs to find the best solution. I did spend a few hours this morning trying to find the "right place" before flattening etc occurs, but need further input.

setchy commented:
Reworked the implementation this morning.  Removed the _crowbar_ approach. Feedback of course welcomed.

Demo repo updated: https://github.com/setchy/renovate-demo-osv-severity/pulls

viceice commented on lib/config/index.spec.ts:
```diff
@@ -109,5 +109,19 @@ describe('config/index', () => {
+      const childConfig = {
+        vulnerabilitySeverity: 'CRITICAL',
+      };
+      const configParser = await import('./index');
```
why new import?

setchy commented on lib/workers/repository/process/vulnerabilities.ts:
```diff
@@ -558,4 +557,32 @@ export class Vulnerabilities {
+    const cvssVector =
+      vulnerability.severity?.find((e) => e.type === 'CVSS_V3')?.score ??
+      vulnerability.severity?.[0]?.score ??
+      (affected.database_specific?.cvss as string); // RUSTSEC
```
Good question - Not too sure @viceice - I kept the existing logic as-is, just needed to move it into a function to make it reusable beyond the`generatePrBodyNotes`

setchy replied:
I looked a bit closer and the type defs for osv-offline-db has this as `unknown`

viceice replied:
ok. @JamieMagee Can we fix that types upstream?

setchy commented on lib/workers/repository/process/vulnerabilities.ts:
```diff
@@ -558,4 +557,32 @@ export class Vulnerabilities {
+      vulnerability.id.startsWith('GHSA-') &&
+      vulnerability.database_specific?.severity
+    ) {
+      const severity = vulnerability.database_specific.severity as string;
```
as above - not too sure honestly.  I kept the existing logic unchanged

setchy replied:
I looked a bit closer and the type defs for osv-offline-db has this as `unknown`

JamieMagee replied:
As above, this is on purpose `database_specific` isn't known ahead of time, and has to be checked at runtime for each vulnerability

setchy commented on lib/util/vulnerability/utils.ts:
```diff
@@ -0,0 +1,31 @@
+export function setHighestVulnerabilitySeverity(
+  config: Record<string, any>,
+  parent: Record<string, any>,
+  child: Record<string, any>
```
Sure - I've updated it to match the types used within the calling function https://github.com/renovatebot/renovate/blob/5b0a0b8bcfb36d7330544d44e6d8dd5afcc0e1fc/lib/config/utils.ts#L5-L8

JamieMagee commented on lib/util/vulnerability/utils.ts:
```diff
@@ -0,0 +1,30 @@
+export function setHighestVulnerabilitySeverity<
+  T extends Record<string, any>,
+  TChild extends Record<string, any> | undefined
+>(config: Record<string, any>, parent: T, child: TChild): void {
```
nit I would make this side-effect free. rename it `getHighestVulnerabilitySeverity` and allow the caller to set `config.vulnerabilitySeverity`.

setchy replied:
Good point - incorporated

JamieMagee commented on lib/util/vulnerability/utils.ts:
```diff
@@ -0,0 +1,30 @@
+
+  const childVulSeverity = child?.vulnerabilitySeverity?.toUpperCase();
+
+  if (!parentVulSeverity && childVulSeverity) {
+    mostSevere = childVulSeverity;
+  } else if (parentVulSeverity && !childVulSeverity) {
+    mostSevere = parentVulSeverity;
+  } else if (parentVulSeverity === 'CRITICAL') {
+    mostSevere = 'CRITICAL';
+  } else if (parentVulSeverity === 'HIGH' && childVulSeverity !== 'CRITICAL') {
+    mostSevere = 'HIGH';
+  } else if (
+    parentVulSeverity === 'MODERATE' &&
+    childVulSeverity !== 'CRITICAL' &&
+    childVulSeverity !== 'HIGH'
+  ) {
+    mostSevere = 'MODERATE';
+  } else {
+    mostSevere = childVulSeverity;
+  }
```
This feels like an awkward way to do this. What about something [like this][1]

[1]: https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAcwKZQBI2QC1QZygDUQAbMVAJwEMAjGUmKATwGVUA3Kp5gHgChEiACqJUADyiowAE3yIASqghxKM3oUowwyADSJqYZgD5dgkQGEcDGWMnS5i5avWbteg0eOIAPonAyqMDaqDL8xgAUAA7UlNJQAFwi+hDWpDJJwlY2AJRJbjqIAN7mKmCEiDFxYMRk7FxaLIgAvJWx8QB0HGQUNPSMLPXcLAD8HVBwAKpRUVQW1PioETkA3KUIFak2JKRDjcwtiFvpY93kVHQMPHs8YxPTs5Tzi8tr6+VQiIsNPADyalQkkoVGoNFAtDp9GAQABbWhUbytEpCIQAGV+AHUkgBGMwoxAAWV+ABEAKIKACCwlJSQATHiURgAJIAcQwSQAzAyhBYFEzhEyLBTUUkACwMgC+byEMGAiAixxkOxuTWaav8siCIRkOWK5iEcSgIEoSCq8WVnGGzDWQgl-HMsvlZpqFp+qvVAS1FB1evxhuNpvaLrqlv2NsQdvM-pNX1DfwBlAA2s7ars4ywALqIYytb5W-6BJOK11WjP6xAjNrVVMq5jlpLFkNu638SNlCp5-aHNCYbB4Qg7XqXAZsdPMCLmIqIM5D-rXMdJABERLJlOpi4jDKnM4uc8GC8Qi+ZbI3dtW9vbcFIqA6pDgyAinZ456AA

setchy replied:
Absolutely, that's way more elegant, thanks @JamieMagee 🙇

secustor commented on lib/util/template/index.ts:
```diff
@@ -143,6 +144,8 @@ export const allowedFields = {
   versioning: 'The versioning scheme in use',
   versions: 'An array of ChangeLogRelease objects in the upgrade',
+  vulnerabilitySeverity:
+    'The vulnerability severity for a vulnerability alert upgrade',
```
We should document which levels are available to the user.

You should be able to dynamically template in the levels.

setchy replied:
Can you elaborate more please @secustor.  

Are you thinking it should mention `LOW, MODERATE, HIGH, CRITICAL` for example?
